addons/product/models/product_template.py

```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
+            record.valid_product_attribute_wnva_ids = record.valid_product_template_attribute_line_wnva_ids.mapped('attribute_id')
+
+            valid_value_ids = record.valid_product_attribute_value_wnva_ids
+            valid_attribute_ids = record.valid_product_attribute_wnva_ids
+
+            record.valid_archived_variant_ids = archived_variants.filtered(
+                lambda v: v.product_tmpl_id == record and v._has_valid_attributes(valid_attribute_ids, valid_value_ids)
+            )
 
     @api.multi
-    def _get_valid_product_attributes(self):
-        """A product attribute is considered valid for a template if it
-        has at least one possible value set on the template.
+    def _compute_valid_existing_variant_ids(self):
+        """This compute is done outside of `_compute_valid_attributes` because
+        it is often not needed, and it can be very bad on performance."""
+        existing_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', True)])
+        for record in self:
 
-        See `_get_valid_product_template_attribute_lines`.
+            valid_value_ids = record.valid_product_attribute_value_wnva_ids
+            valid_attribute_ids = record.valid_product_attribute_wnva_ids
 
-        :return: all the valid product attributes of this template
-        :rtype: recordset `product.attribute`
-        """
+            record.valid_existing_variant_ids = existing_variants.filtered(
+                lambda v: v.product_tmpl_id == record and v._has_valid_attributes(valid_attribute_ids, valid_value_ids)
+            )
+
+    @api.multi
+    def _get_valid_product_template_attribute_lines(self):
+        """deprecated, use `valid_product_template_attribute_line_ids`"""
         self.ensure_one()
-        product_attributes = self.env['product.attribute']
-        for ptal in self._get_valid_product_template_attribute_lines():
-            product_attributes |= ptal.attribute_id
-        return product_attributes
+        return self.valid_product_template_attribute_line_ids
 
     @api.multi
-    def _get_valid_product_attribute_values(self):
-        """A product attribute value is considered valid for a template if it is
-        defined on a product template attribute line.
+    def _get_valid_product_attributes(self):
+        """deprecated, use `valid_product_attribute_ids`"""
+        self.ensure_one()
+        return self.valid_product_attribute_ids
 
-        :return: all the valid product attribute values of this template
-        :rtype: recordset `product.attribute.value`
-        """
+    @api.multi
+    def _get_valid_product_attribute_values(self):
+        """deprecated, use `valid_product_attribute_value_ids`"""
         self.ensure_one()
-        return self._get_valid_product_template_attribute_lines().mapped('value_ids')
+        return self.valid_product_attribute_value_ids
```

Comment:
add logger warning

Reply:
I didn't actually change the existing code using the deprecated methods to keep a diff minimal.

It's not a big deal to use it, since the method is now using the computed field, it's performant too, it's just a duplicate.